src/Uno.UI/UI/Xaml/Controls/Popup/Popup.Base.cs

```diff
@@ -154,26 +157,23 @@ protected internal override void OnTemplatedParentChanged(DependencyPropertyChan
 
 	private void UpdateDataContext(DependencyPropertyChangedEventArgs e)
 	{
-		_childHasOwnDataContext = false;
 		if (PropagatesDataContextToChild && Child is IDependencyObjectStoreProvider provider)
 		{
-			var dataContextProperty = provider.Store.ReadLocalValue(provider.Store.DataContextProperty);
+			var childLocalValue = provider.Store.ReadLocalValue(provider.Store.DataContextProperty);
+			_childHasOwnDataContext = childLocalValue != null && childLocalValue != DependencyProperty.UnsetValue;
+
+			var childPopupValue = provider.Store.GetValue(provider.Store.DataContextProperty, DependencyPropertyValuePrecedences.Popup);
 
-			var shouldClearValue = e != null && e.NewValue == null && dataContextProperty == e.OldValue;
+			var shouldClearValue = e != null && e.NewValue == null && childPopupValue == e.OldValue;
 			if (shouldClearValue)
 			{
 				//In this case we are clearing the DataContext that was previously set by the Popup
 				//This usually occurs when the owner of the Popup is removed from the Visual Tree
-				provider.Store.ClearValue(provider.Store.DataContextProperty, DependencyPropertyValuePrecedences.Local);
-			}
-			else if (dataContextProperty != null && dataContextProperty != DependencyProperty.UnsetValue)
-			{
-				// Child already has locally set DataContext, we shouldn't overwrite it.
-				_childHasOwnDataContext = true;
+				provider.Store.ClearValue(provider.Store.DataContextProperty, DependencyPropertyValuePrecedences.Popup);
 			}
 			else
 			{
-				provider.Store.SetValue(provider.Store.DataContextProperty, this.DataContext, DependencyPropertyValuePrecedences.Local);
+				provider.Store.SetValue(provider.Store.DataContextProperty, this.DataContext, DependencyPropertyValuePrecedences.Popup);
```

Comment:
When the issue happens, at which precedence is the offending datacontext being set? In some cases, the datacontext may be incorrectly set at the style level, because of the popup being created during a style for instance.

Reply:
I'm not sure what you mean by the "offending" data context. The problem is that the popup sets the DataContext for its logical child at the Local level, which causes an ambiguity between the DataContext set by the popup and an actual Local Datacontext. The other obvious option is to set the DataContext at the inheritance level, except that the the logical child's visual parent (mostly PopupRoot) will override the Inheritance DataContext (to null).

Reply:
Then we may need to break that link from the `PopupRoot`, and prevent the DataContext to change. Depending on how the parent is set, we may able to use this:
https://github.com/unoplatform/uno/blob/06ebb2b56d9df2fb0b35be688fd9900483efd870/src/Uno.UI/UI/Xaml/FrameworkPropertyMetadataOptions.cs#L35

Reply:
That didn't work out, I'm not exactly sure why. @Youssef1313 had a clever idea to set the DataContext on `PopupPanel` instead of the child itself (and then the child can inherit those values). `PopupPanel` is internal, so we don't need to adhere to what happens on WinUI. I'm not sure if this has any unintended side effects. Let's see what CI says.

Reply:
Looks like it did help! Very nice.